datafusion/core/src/physical_optimizer/sort_enforcement.rs

```diff
@@ -33,48 +33,80 @@ use crate::config::ConfigOptions;
 use crate::error::Result;
 use crate::physical_optimizer::utils::add_sort_above_child;
 use crate::physical_optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::coalesce_partitions::CoalescePartitionsExec;
+use crate::physical_plan::limit::{GlobalLimitExec, LocalLimitExec};
 use crate::physical_plan::rewrite::TreeNodeRewritable;
 use crate::physical_plan::sorts::sort::SortExec;
+use crate::physical_plan::sorts::sort_preserving_merge::SortPreservingMergeExec;
+use crate::physical_plan::union::UnionExec;
 use crate::physical_plan::windows::{BoundedWindowAggExec, WindowAggExec};
-use crate::physical_plan::{with_new_children_if_necessary, ExecutionPlan};
+use crate::physical_plan::{with_new_children_if_necessary, Distribution, ExecutionPlan};
 use arrow::datatypes::SchemaRef;
 use datafusion_common::{reverse_sort_options, DataFusionError};
 use datafusion_physical_expr::utils::{ordering_satisfy, ordering_satisfy_concrete};
-use datafusion_physical_expr::window::WindowExpr;
 use datafusion_physical_expr::{PhysicalExpr, PhysicalSortExpr};
-use itertools::izip;
+use itertools::{concat, izip};
 use std::iter::zip;
 use std::sync::Arc;
 
 /// This rule inspects SortExec's in the given physical plan and removes the
-/// ones it can prove unnecessary.
+/// ones it can prove unnecessary. The boolean flag `parallelize_sorts`
+/// indicates whether we elect to transform CoalescePartitionsExec + SortExec
+/// cascades into SortExec + SortPreservingMergeExec cascades, which enables
+/// us to perform sorting in parallel.
 #[derive(Default)]
-pub struct EnforceSorting {}
+pub struct EnforceSorting {
+    parallelize_sorts: bool,
+}
 
 impl EnforceSorting {
     #[allow(missing_docs)]
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(parallelize_sorts: bool) -> Self {
```

Comment:
I didn't see any place in the code that ever crated an `EnforceSorting` with `parallelize_sorts = false` (aka I searched for `EnforceSorting::new(false)`. I may have missed it. 

I think we should either remove the code that handles parallelize_sorts = false as it is not used/tested or better yet (as explained in the overall PR comments) that we add a new config option to control the behavior and add some basic tests.

Reply:
We will modify the tests to verify both cases, thank you for pointing it out. I also agree with making this flag a configuration parameter, will do that too.

Reply:
This flag is now a configuration parameter. I add a test to verify `false` case also. Test can be found [here](https://github.com/synnada-ai/arrow-datafusion/blob/0ddc0a756a032081bee2f8f29995c35b50d72b6a/datafusion/core/tests/sql/window.rs#L2426-L2462)